zerver/webhooks/slack/view.py

```diff
@@ -1,60 +1,238 @@
 @typed_endpoint
 def api_slack_webhook(
     request: HttpRequest,
     user_profile: UserProfile,
     *,
-    user_name: str,
-    text: str,
-    channel_name: str,
+    payload: JsonBodyPayload[WildValue],
+    slack_app_token: str,
     channels_map_to_topics: str | None = None,
 ) -> HttpResponse:
-    content = ZULIP_MESSAGE_TEMPLATE.format(message_sender=user_name, text=text)
     topic_name = "Message from Slack"
 
-    if channels_map_to_topics is None:
+    # Handle initial URL verification handshake for Slack Events API.
+    if is_challenge_handshake(payload):
+        challenge = payload.get("challenge").tame(check_string)
         check_send_webhook_message(
             request,
             user_profile,
-            topic_name,
-            content,
-        )
-    elif channels_map_to_topics == VALID_OPTIONS["SHOULD_BE_MAPPED"]:
-        # If the webhook URL has a user_specified_topic,
-        # then this topic-channel mapping will not be used.
-        topic_name = f"channel: {channel_name}"
-        check_send_webhook_message(
-            request,
-            user_profile,
-            topic_name,
-            content,
-        )
-    elif channels_map_to_topics == VALID_OPTIONS["SHOULD_NOT_BE_MAPPED"]:
-        # This channel-channel mapping will be used even if
-        # there is a channel specified in the webhook URL.
-        check_send_webhook_message(
-            request,
-            user_profile,
-            topic_name,
-            content,
-            stream=channel_name,
+            "Integration events",
+            "Successfully verified webhook URL with Slack!",
         )
+        return json_success(request=request, data={"challenge": challenge})
+
+    if is_retry_call_from_slack(request):
+        # A Slack fail condition occurs when we don't respond with HTTP 200
+        # within 3 seconds after Slack calls our endpoint. If this happens,
+        # Slack will retry sending the same payload. This is often triggered
+        # because of we have to do two callbacks for each call. To avoid
+        # sending the same message multiple times, we block subsequent retry
+        # calls from Slack.
+        return json_success(request)
+
+    # Prevent any Zulip messages sent through the Slack Bridge from looping
+    # back here.
+    if is_zulip_slack_bridge_bot_message(payload):
+        return json_success(request)
+
+    check_token_access(slack_app_token)
+
+    event_dict = payload.get("event", {})
+    event_type = event_dict.get("type").tame(check_string)
+    if event_type == "message":
+        file_dict = event_dict.get("files")
+        raw_text = event_dict.get("text", "").tame(check_string)
+
+        event_data: dict[str, Any] = {
+            "type": event_dict.get("type").tame(check_string),
+            "text": convert_to_zulip_markdown(raw_text, slack_app_token),
+            "channel": get_channel_name_str(payload, slack_app_token),
```

Comment:
I wonder if we want to avoid this callback unless there's some sort of valid `channels_map_to_topics` configuration?

```python
if channels_map_to_topics is None:
    ...
elif channels_map_to_topics in VALID_OPTIONS:
    channel = get_channel_name_str ...
    if channels_map_to_topics == VALID_OPTIONS["SHOULD_BE_MAPPED"]:
        ...
    else:
        ...
else:
    JsonableError ...
```

Reply:
Right, this is a good catch.